File: framec_tests/golang/hierarchical/hierarchical.go
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files


func NewHierarchical() Hierarchical {
    m := &hierarchicalStruct{}
    
    // Validate interfaces
    var _ Hierarchical = m
    var _ Hierarchical_actions = m
    
    // Create and intialize start state compartment.
    m._compartment_ = NewHierarchicalCompartment(HierarchicalState_I)
    
    // Override domain variables.
    m.enters = []string{}
    m.exits = []string{}
    m.tape = []string{}
    
    // Send system start event
    e := framelang.FrameEvent{Msg:">"}
    m._mux_(&e)
    return m
}


type HierarchicalState uint

const (
    HierarchicalState_I HierarchicalState = iota
    HierarchicalState_S
    HierarchicalState_S0
    HierarchicalState_S1
    HierarchicalState_S2
    HierarchicalState_S3
    HierarchicalState_T
)

type Hierarchical interface {
    A() 
    B() 
    C() 
}

type Hierarchical_actions interface {
    enter(msg string) 
    exit(msg string) 
    log(msg string) 
}


type hierarchicalStruct struct {
    _compartment_ *HierarchicalCompartment
    _nextCompartment_ *HierarchicalCompartment
    enters []string
    exits []string
    tape []string
}

//===================== Interface Block ===================//

func (m *hierarchicalStruct) A()  {
    e := framelang.FrameEvent{Msg:"A"}
    m._mux_(&e)
}

func (m *hierarchicalStruct) B()  {
    e := framelang.FrameEvent{Msg:"B"}
    m._mux_(&e)
}

func (m *hierarchicalStruct) C()  {
    e := framelang.FrameEvent{Msg:"C"}
    m._mux_(&e)
}

//====================== Multiplexer ====================//

func (m *hierarchicalStruct) _mux_(e *framelang.FrameEvent) {
    switch m._compartment_.State {
    case HierarchicalState_I:
        m._HierarchicalState_I_(e)
    case HierarchicalState_S:
        m._HierarchicalState_S_(e)
    case HierarchicalState_S0:
        m._HierarchicalState_S0_(e)
    case HierarchicalState_S1:
        m._HierarchicalState_S1_(e)
    case HierarchicalState_S2:
        m._HierarchicalState_S2_(e)
    case HierarchicalState_S3:
        m._HierarchicalState_S3_(e)
    case HierarchicalState_T:
        m._HierarchicalState_T_(e)
    }
    
    if m._nextCompartment_ != nil {
        nextCompartment := m._nextCompartment_
        m._nextCompartment_ = nil
        if nextCompartment._forwardEvent_ != nil && 
           nextCompartment._forwardEvent_.Msg == ">" {
            m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
            m._compartment_ = nextCompartment
            m._mux_(nextCompartment._forwardEvent_)
        } else {
            m._do_transition_(nextCompartment)
            if nextCompartment._forwardEvent_ != nil {
                m._mux_(nextCompartment._forwardEvent_)
            }
        }
        nextCompartment._forwardEvent_ = nil
    }
}

//===================== Machine Block ===================//

func (m *hierarchicalStruct) _HierarchicalState_I_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        compartment := NewHierarchicalCompartment(HierarchicalState_S)
        m._transition_(compartment)
        
        return
    }
}

func (m *hierarchicalStruct) _HierarchicalState_S_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.enter("S")
        
        return
    case "<":
        m.exit("S")
        
        return
    case "A":
        m.log("S.A")
        compartment := NewHierarchicalCompartment(HierarchicalState_S0)
        m._transition_(compartment)
        
        return
    case "B":
        m.log("S.B")
        compartment := NewHierarchicalCompartment(HierarchicalState_S1)
        m._transition_(compartment)
        
        return
    }
}

func (m *hierarchicalStruct) _HierarchicalState_S0_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.enter("S0")
        
    case "<":
        m.exit("S0")
        
      //  override parent handler
    case "A":
        m.log("S0.A")
        compartment := NewHierarchicalCompartment(HierarchicalState_T)
        m._transition_(compartment)
        
        return
      //  do this, then parent handler
    case "B":
        m.log("S0.B")
        
      //  extend parent handler
    case "C":
        m.log("S0.C")
        compartment := NewHierarchicalCompartment(HierarchicalState_S2)
        m._transition_(compartment)
        
        return
    }
    m._HierarchicalState_S_(e)
    
}

func (m *hierarchicalStruct) _HierarchicalState_S1_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.enter("S1")
        
        return
    case "<":
        m.exit("S1")
        
        return
      //  defer to parent for A
      //  do this, then parent, which transitions here
    case "B":
        m.log("S1.B")
        
      //  propagate message not handled by parent
    case "C":
        m.log("S1.C")
        
    }
    m._HierarchicalState_S_(e)
    
}

func (m *hierarchicalStruct) _HierarchicalState_S2_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.enter("S2")
        
    case "<":
        m.exit("S2")
        
      //  will propagate to S0 and S
    case "B":
        m.log("S2.B")
        
    case "C":
        m.log("S2.C")
        compartment := NewHierarchicalCompartment(HierarchicalState_T)
        m._transition_(compartment)
        
        return
    }
    m._HierarchicalState_S0_(e)
    
}  //  continue after transition (should be ignored)


func (m *hierarchicalStruct) _HierarchicalState_S3_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.enter("S3")
        
    case "<":
        m.exit("S3")
        
      //  defer to grandparent for A
      //  override and move to sibling
    case "B":
        m.log("S3.B")
        compartment := NewHierarchicalCompartment(HierarchicalState_S2)
        m._transition_(compartment)
        
        return
    }
    m._HierarchicalState_S1_(e)
    
}

func (m *hierarchicalStruct) _HierarchicalState_T_(e *framelang.FrameEvent) {
    switch e.Msg {
    case ">":
        m.enter("T")
        
        return
    case "<":
        m.exit("T")
        
        return
    case "A":
        m.log("T.A")
        compartment := NewHierarchicalCompartment(HierarchicalState_S)
        m._transition_(compartment)
        
        return
    case "B":
        m.log("T.B")
        compartment := NewHierarchicalCompartment(HierarchicalState_S2)
        m._transition_(compartment)
        
        return
    case "C":
        m.log("T.C")
        compartment := NewHierarchicalCompartment(HierarchicalState_S3)
        m._transition_(compartment)
        
        return
    }
}

//=============== Machinery and Mechanisms ==============//

func (m *hierarchicalStruct) _transition_(compartment *HierarchicalCompartment) {
    m._nextCompartment_ = compartment
}

func (m *hierarchicalStruct) _do_transition_(nextCompartment *HierarchicalCompartment) {
    m._mux_(&framelang.FrameEvent{Msg: "<", Params: m._compartment_.ExitArgs, Ret: nil})
    m._compartment_ = nextCompartment
    m._mux_(&framelang.FrameEvent{Msg: ">", Params: m._compartment_.EnterArgs, Ret: nil})
}

//===================== Actions Block ===================//


/********************************************************

// Unimplemented Actions

func (m *hierarchicalStruct) enter(msg string)  {}
func (m *hierarchicalStruct) exit(msg string)  {}
func (m *hierarchicalStruct) log(msg string)  {}

********************************************************/

//=============== Compartment ==============//

type HierarchicalCompartment struct {
    State HierarchicalState
    StateArgs map[string]interface{}
    StateVars map[string]interface{}
    EnterArgs map[string]interface{}
    ExitArgs map[string]interface{}
    _forwardEvent_ *framelang.FrameEvent
}

func NewHierarchicalCompartment(state HierarchicalState) *HierarchicalCompartment {
    c := &HierarchicalCompartment{State: state}
    c.StateArgs = make(map[string]interface{})
    c.StateVars = make(map[string]interface{})
    c.EnterArgs = make(map[string]interface{})
    c.ExitArgs = make(map[string]interface{})
    return c
}

File: framec_tests/python/src/hierarchical/hierarchical.py
#Emitted from framec_v0.11.2



from framelang.framelang import FrameEvent



class FrameEvent:
    def __init__(self, message, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class Hierarchical:
    
    
    # ==================== System Factory =================== #
    
    def __init__(self):
        
         # Create and initialize start state compartment.
        
        next_compartment = None
        next_compartment = HierarchicalCompartment('__hierarchical_state_I', next_compartment)
        self.__compartment = next_compartment
        self.__next_compartment = None
        self.return_stack = [None]
        
        # Initialize domain
        
        self.enters  = []
        self.exits  = []
        self.tape  = []
        
        # Send system start event
        frame_event = FrameEvent(">", None)
        self.__kernel(frame_event)
    
    # ==================== Interface Block ================== #
    
    def A(self,):
        self.return_stack.append(None)
        __e = FrameEvent("A",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def B(self,):
        self.return_stack.append(None)
        __e = FrameEvent("B",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    def C(self,):
        self.return_stack.append(None)
        __e = FrameEvent("C",None)
        self.__kernel(__e)
        self.return_stack.pop(-1)
    
    # ===================== Machine Block =================== #
    
    # ----------------------------------------
    # $I
    
    def __hierarchical_state_I(self, __e, compartment):
        if __e._message == ">":
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_S', next_compartment)
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $S
    
    def __hierarchical_state_S(self, __e, compartment):
        if __e._message == ">":
            self.enter_do("S")
            return
        elif __e._message == "<":
            self.exit_do("S")
            return
        elif __e._message == "A":
            self.log_do("S.A")
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_S', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S0', next_compartment)
            self.__transition(next_compartment)
            return
        elif __e._message == "B":
            self.log_do("S.B")
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_S', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S1', next_compartment)
            self.__transition(next_compartment)
            return
    
    # ----------------------------------------
    # $S0
    
    def __hierarchical_state_S0(self, __e, compartment):
        if __e._message == ">":
            self.enter_do("S0")
        elif __e._message == "<":
            self.exit_do("S0")
          #  override parent handler
        elif __e._message == "A":
            self.log_do("S0.A")
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_T', next_compartment)
            self.__transition(next_compartment)
            return
          #  do this, then parent handler
        elif __e._message == "B":
            self.log_do("S0.B")
          #  extend parent handler
        elif __e._message == "C":
            self.log_do("S0.C")
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_S', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S0', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S2', next_compartment)
            self.__transition(next_compartment)
            return
        
        self.__hierarchical_state_S(__e, compartment.parent_compartment)
        
    
    # ----------------------------------------
    # $S1
    
    def __hierarchical_state_S1(self, __e, compartment):
        if __e._message == ">":
            self.enter_do("S1")
            return
        elif __e._message == "<":
            self.exit_do("S1")
            return
          #  defer to parent for A
          #  do this, then parent, which transitions here
        elif __e._message == "B":
            self.log_do("S1.B")
          #  propagate message not handled by parent
        elif __e._message == "C":
            self.log_do("S1.C")
        
        self.__hierarchical_state_S(__e, compartment.parent_compartment)
        
    
    # ----------------------------------------
    # $S2
    
    def __hierarchical_state_S2(self, __e, compartment):
        if __e._message == ">":
            self.enter_do("S2")
        elif __e._message == "<":
            self.exit_do("S2")
          #  will propagate to S0 and S
        elif __e._message == "B":
            self.log_do("S2.B")
        elif __e._message == "C":
            self.log_do("S2.C")
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_T', next_compartment)
            self.__transition(next_compartment)
            return
        
        self.__hierarchical_state_S0(__e, compartment.parent_compartment)
        
      #  continue after transition (should be ignored)
    
    
    # ----------------------------------------
    # $S3
    
    def __hierarchical_state_S3(self, __e, compartment):
        if __e._message == ">":
            self.enter_do("S3")
        elif __e._message == "<":
            self.exit_do("S3")
          #  defer to grandparent for A
          #  override and move to sibling
        elif __e._message == "B":
            self.log_do("S3.B")
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_S', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S0', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S2', next_compartment)
            self.__transition(next_compartment)
            return
        
        self.__hierarchical_state_S1(__e, compartment.parent_compartment)
        
    
    # ----------------------------------------
    # $T
    
    def __hierarchical_state_T(self, __e, compartment):
        if __e._message == ">":
            self.enter_do("T")
            return
        elif __e._message == "<":
            self.exit_do("T")
            return
        elif __e._message == "A":
            self.log_do("T.A")
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_S', next_compartment)
            self.__transition(next_compartment)
            return
        elif __e._message == "B":
            self.log_do("T.B")
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_S', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S0', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S2', next_compartment)
            self.__transition(next_compartment)
            return
        elif __e._message == "C":
            self.log_do("T.C")
            next_compartment = None
            next_compartment = HierarchicalCompartment('__hierarchical_state_S', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S1', next_compartment)
            next_compartment = HierarchicalCompartment('__hierarchical_state_S3', next_compartment)
            self.__transition(next_compartment)
            return
    
    # ===================== Actions Block =================== #
    
    def enter_do(self,msg: str):
        raise NotImplementedError
    
    def exit_do(self,msg: str):
        raise NotImplementedError
    
    def log_do(self,msg: str):
        raise NotImplementedError
    
    # ==================== System Runtime =================== #
    
    def __kernel(self, __e):
        
        # send event to current state
        self.__router(__e)
        
        # loop until no transitions occur
        while self.__next_compartment != None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            
            # exit current state
            self.__router(FrameEvent( "<", self.__compartment.exit_args))
            # change state
            self.__compartment = next_compartment
            
            if next_compartment.forward_event is None:
                # send normal enter event
                self.__router(FrameEvent(">", self.__compartment.enter_args))
            else: # there is a forwarded event
                if next_compartment.forward_event._message == ">":
                    # forwarded event is enter event
                    self.__router(next_compartment.forward_event)
                else:
                    # forwarded event is not enter event
                    # send normal enter event
                    self.__router(FrameEvent(">", self.__compartment.enter_args))
                    # and now forward event to new, intialized state
                    self.__router(next_compartment.forward_event)
                next_compartment.forward_event = None
                
    
    def __router(self, __e):
        if self.__compartment.state == '__hierarchical_state_I':
            self.__hierarchical_state_I(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchical_state_S':
            self.__hierarchical_state_S(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchical_state_S0':
            self.__hierarchical_state_S0(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchical_state_S1':
            self.__hierarchical_state_S1(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchical_state_S2':
            self.__hierarchical_state_S2(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchical_state_S3':
            self.__hierarchical_state_S3(__e, self.__compartment)
        elif self.__compartment.state == '__hierarchical_state_T':
            self.__hierarchical_state_T(__e, self.__compartment)
        
    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment
    
    def state_info(self):
        return self.__compartment.state
        

# ===================== Compartment =================== #

class HierarchicalCompartment:

    def __init__(self,state,parent_compartment):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment
    